File: src/dexter/frontend/clang/src/ir/bool/FuncDecl.cpp
/*
 * FuncDecl.cpp
 *
 *  Created on: Mar 24, 2018
 *      Author: maazsaf
 */

#include "ir/bool/FuncDecl.h"

Dexter::FuncDecl::FuncDecl (std::string name, std::vector<VarExpr *> params, Dexter::Type retType, Dexter::Expr * body, bool isGen) : Dexter::Expr(NULL)
{
  JNIEnv * e = Util::env;

  jclass c = e->FindClass("java/util/ArrayList");
  jmethodID m = e->GetMethodID(c, "<init>", "()V");
  jobject list = e->NewObject(c, m);

  m = e->GetMethodID(c, "add", "(Ljava/lang/Object;)Z");
  std::vector<VarExpr *>::const_iterator sit;
  for (sit = params.begin(); sit != params.end(); ++sit)
    e->CallObjectMethod(list, m, (*sit)->obj());

  jstring n = e->NewStringUTF(name.c_str());

	c = e->FindClass("dexter/ir/bool/FuncDecl");
	m = e->GetMethodID(c, "<init>", "(Ljava/lang/String;Ljava/util/List;Ldexter/ir/type/Type;Ldexter/ir/Expr;Z)V");
	this->mObj = e->NewObject(c, m, n, list, retType.obj(), body->obj(), isGen);
}

File: src/dexter/ir/bool/FuncDecl.java
package dexter.ir.bool;

import dexter.ir.Expr;
import dexter.ir.Visitor;
import dexter.ir.type.Type;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Maaz Ahmad on 6/25/19.
 */
public class FuncDecl extends Expr
{
  protected String name;
  protected List<VarExpr> params;
  protected Expr body;
  protected Type retType;
  protected boolean isUnInterpreted;
  protected boolean isGenerator;

  public FuncDecl (String name, List<VarExpr> params, Type retType, Expr body, boolean isGen)
  {
    this.name = name;
    this.params = new ArrayList<> (params);
    this.body = body;
    this.isUnInterpreted = (body == null ? true : false);
    this.retType = retType;
    this.isGenerator = isGen;
  }

  public String name () { return name; }
  public List<VarExpr> params () { return params; }
  public Expr body () { return body; }
  public Type retType () { return retType; }
  public boolean isUnInterpreted () { return isUnInterpreted; }
  public void isUnInterpreted (boolean isU) { isUnInterpreted = isU; }
  public boolean isGenerator () { return isGenerator; }
  public void isGenerator (boolean isG) { isGenerator = isG; }

  public void body (Expr body) { this.body = body; }
  public void params (List<VarExpr> params) { this.params = params; }

  @Override
  public <T> T accept(Visitor<T> v)
  {
    return v.visit(this);
  }

  @Override
  public boolean equals (Object o)
  {
    if (o instanceof FuncDecl)
    {
      FuncDecl other = (FuncDecl)o;
      return other.name.equals(this.name) && other.params.equals(this.params) &&
              other.isUnInterpreted == this.isUnInterpreted && (other.isUnInterpreted || other.body.equals(this.body));
    }
    else
      return false;
  }
}